pangeo_forge_recipes/transforms.py

```diff
@@ -688,31 +667,51 @@ def expand(
         schema = datasets | DetermineSchema(combine_dims=self.combine_dims)
-        indexed_datasets = datasets | IndexItems(schema=schema, append_offset=self._append_offset)
-        target_chunks = (
-            self.target_chunks
+
+        # Index datasets according to their place within the global schema
+        indexed_datasets = datasets | beam.ParDo(
+            IndexWithPosition(append_offset=self._append_offset),
+            schema=beam.pvalue.AsSingleton(schema),
+        )
+
+        # either use target chunks or else compute them with provided chunking function
+        # Make a PColl for target chunks to match the output of mapping a dynamic chunking fn
+        target_chunks_pcoll = (
+            pipeline | "Create target_chunks pcoll" >> beam.Create([self.target_chunks])
             if not self.dynamic_chunking_fn
-            else beam.pvalue.AsSingleton(
+            else (
                 schema
-                | beam.Map(schema_to_template_ds)
-                | beam.Map(self.dynamic_chunking_fn, **self.dynamic_chunking_fn_kwargs)
+                | "make template dataset" >> beam.Map(schema_to_template_ds)
+                | "generate chunks dynamically" >> beam.Map(self.dynamic_chunking_fn)
             )
         )
-        logger.info(f"Storing Zarr with {target_chunks =} to {self.get_full_target()}")
-        rechunked_datasets = indexed_datasets | Rechunk(target_chunks=target_chunks, schema=schema)
-        target_store = schema | PrepareZarrTarget(
+
+        # split datasets according to their write-targets (chunks of bytes)
+        # then combine datasets with shared write targets
+        # Note that the pipe (|) operator in beam is just sugar for passing into expand
+        rechunked_datasets = ChunkToTarget().expand(
+            indexed_datasets,
+            target_chunks=beam.pvalue.AsSingleton(target_chunks_pcoll),
+            schema=beam.pvalue.AsSingleton(schema),
+        )
+
+        target_store = PrepareZarrTarget(
             target=self.get_full_target(),
-            target_chunks=target_chunks,
             attrs=self.attrs,
             encoding=self.encoding,
             append_dim=self.append_dim,
-        )
-        n_target_stores = rechunked_datasets | StoreDatasetFragments(target_store=target_store)
-        singleton_target_store = (
-            n_target_stores
-            | beam.combiners.Sample.FixedSizeGlobally(1)
-            | beam.FlatMap(lambda x: x)  # https://stackoverflow.com/a/47146582
+        ).expand(schema, beam.pvalue.AsSingleton(target_chunks_pcoll))
+
+        # Actually attempt to write datasets to their target bytes/files
+        rechunked_datasets | "write chunks" >> beam.Map(
+            store_dataset_fragment, target_store=beam.pvalue.AsSingleton(target_store)
         )
 
-        return singleton_target_store
+        # return the target store pcollection (a singleton of the fsspec target)
+        return target_store
```

Comment:
The reason we did this `n_target_stores` -> `FixedSizeGlobally(1)` nonsense before (and sorry for not documenting this in comments or otherwise!) ... was that without doing this, Beam will not wait on all chunks to be finished writing before yielding back the target store... and in this case, I'm pretty sure we do want to enforce blocking behavior because if we don't, then if we try to do:

```python
... | StoreToZarr() | ConsolidateMetadata()
```

```python
... | StoreToZarr() | ValidateStoredZarr()
```

or what-have-you, then the transform that is chained onto `StoreToZarr` will receive it's input before chunks have finished writing. Which will make it impossible to consolidate or validate, etc.

So basically, `n_target_stores` -> `FixedSizeGlobally(1)` ensures we block, and don't let the output of `StoreToZarr` be emitted downstream until all chunks are written.

Does that make sense?

Reply:
Great catch! Makes total sense